### PR TITLE
AC-544: Re-developing OpenMRSLogger

### PR DESCRIPTION
**Pull request by deepak140596**
## Description of what I changed

- Didn't use any other Logging libraries as the OpenMRSLogger had all the functions that was needed for logging. 
- Added an AsyncTask for every Write operation to the Log file on the disk.
- Added an extra method for custom message logging that can be used to format the log file. It will help in future tasks such as developing the Analytics in the app 

## Issue I worked on

JIRA Issue: https://issues.openmrs.org/browse/AC-544

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

- [ ] I have **added tests** to cover my changes. (If you refactored
existing code that was well tested you do not have to add tests)

- [ ] All new and existing **tests passed**.

- [x] My pull request is **based on the latest changes** of the master branch.

### CODE REVIEW

**Comment by Kapil706** [comment]
please correct the pull request template the x needs to be fitted clearly to get the symbol

**Comment by deepak140596** [comment]
@Kapil706 , thanks for pointing out the error.

**Comment by deepak140596** [comment]
Hi @f4ww4z , please review this PR. It results in a lot of performance incerase in the app.

File: openmrs-client/src/main/java/org/openmrs/mobile/application/OpenMRSLogger.java [hunk 1]
```diff
@@ -87,11 +88,36 @@ private static void setErrorCount() {
...
+        // doesn't need any parameter
+        asyncTask.execute();
+    }
+
+    /*
+    This will write custom messages to the logging file
+    The messages will not appear in the logcat window
+    For displaying messages in logcat window, already Log.i,d,w ,etc are defined
+    This can be used for formatting the log file as where necessary and adding important data for analytics
+     */
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/application/OpenMRSLogger.java]: Can you follow the javadoc standard to document this method? Make sure to describe the parameter as well.

File: openmrs-client/src/main/java/org/openmrs/mobile/application/OpenMRSLogger.java [hunk 2]
```diff
@@ -87,11 +88,36 @@ private static void setErrorCount() {
...
+        WriteMsgToFile asyncTask = new WriteMsgToFile();
+        // pass the custom message for execution
+        asyncTask.execute(msg);
+    }
+
+    /*
+    This function writes logs to file
+    The input parameter is empty
+     */
+    private static void saveToFileAsync(){
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/application/OpenMRSLogger.java]: Format the file with Android Studio's formatter.

File: openmrs-client/src/main/java/org/openmrs/mobile/application/OpenMRSLogger.java [hunk 3]
```diff
@@ -115,7 +141,46 @@ private static void saveToFile() {
...
                 if (isSaveToFileEnable()) {
-                    logger.e("Error during waitng for \"logcat -c\" process", e);
+                    logger.e("Error during waiting for \"logcat -c\" process", e);
+                }
+            }
+            rotateLogFile();
+        }
+    }
+
+    private static  class SaveAsyncToFile extends AsyncTask<Void,Void,Void>{
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/application/OpenMRSLogger.java]: Rename this class to `SaveToFileAsyncTask` , and just move the content of the `saveToFileAsync` method here, to make it more concise. Remove the `saveToFileAsync` method. Remember to document the class using javadoc format.

File: openmrs-client/src/main/java/org/openmrs/mobile/application/OpenMRSLogger.java [hunk 4]
```diff
@@ -115,7 +141,46 @@ private static void saveToFile() {
...
+            return null;
+        }
+    }
+
+    private static class WriteMsgToFile extends AsyncTask<String,Void,Void>{
+        @Override
+        // custom message as string parameter
+        protected Void doInBackground(String... strings) {
+            String msg = strings[0];
+            writeMsgToFileAsync(msg);
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/application/OpenMRSLogger.java]: Same comment as for the previous class. Move the content of the `writeMsgToFileAsync` method here, and remove that method, to make it more readable. Describe the class using javadoc format.

File: openmrs-client/src/main/java/org/openmrs/mobile/application/OpenMRSLogger.java [hunk 5]
```diff
@@ -115,7 +141,46 @@ private static void saveToFile() {
...
+                FileWriter writer = new FileWriter(mLogFile, true);
+                writer.write(msg + "\n");
+
+                writer.flush();
+                writer.close();
+
+            } catch (IOException e) {
+                setErrorCount();
+                if (isSaveToFileEnable()) {
+                    logger.e("Error during save log to file", e);
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/application/OpenMRSLogger.java]:
```suggestion
                    logger.e("Error while saving log: ", e);
```

**Pull request closed without merging**